source/agora/consensus/EnrollmentManager.d

```diff
@@ -682,6 +639,56 @@ public class EnrollmentManager
         }
     }
 
+    /***************************************************************************
+
+        Generate pre-images
+
+        Params:
+            current_height = block height used to determine which range of
+                pre-images will be generated.
+
+    ***************************************************************************/
+
+    private void generatePreimages(ulong current_height) @safe nothrow
+    {
+        // if there is no pre-image, the defined number of pre-images must
+        // be generated with the interval of ValidatorCycle.
+        if (this.preimages.length == 0)
+        {
+            auto hash = hashFull(node_key_pair.secret);
+            uint current_idx = AllCountPrimages;
+            this.preimages[current_idx] = hash;
+            foreach (ulong i; 0 .. AllCountPrimages)
+            {
+                hash = hashFull(hash);
+                if (--current_idx % ValidatorSet.ValidatorCycle == 0)
+                    this.preimages[current_idx] = hash;
+            }
+        }
+
+        // This determines which range of preimages must be generated.
+        ulong start_index =
+            ((current_height / ValidatorSet.ValidatorCycle) + 2) *
```

Comment:
Why is there a `+2` here?

Reply:
For example, if the start height is 1800, it needs the 3024th hash value of the `this.preimages`. For detail, 1800 / 1008 is 1, and the value + 2 is 3,  and 3 * 1008 is 3024. That's why. Thank you.

Reply:
In the current design if the Validator wants to start enrolling at block height `1_000_000`, it means that 1 million hashes will be discarded after being generated. Because the preimage to be used is currently tied to the block height.

Reply:
> In the current design if the Validator wants to start enrolling at block height `1_000_000`, it means that 1 million hashes will be discarded after being generated. Because the preimage to be used is currently tied to the block height.

You're right. I will change the line `foreach (ulong i; 0 .. AllCountPrimages)` to a line like `foreach (ulong i; current_height-1 .. AllCountPrimages)'